common/src/main/java/com/microsoft/identity/common/internal/controllers/LocalMSALController.java

```diff
@@ -316,25 +318,31 @@ public AcquireTokenResult acquireTokenSilent(
         // subsequent CacheRecords represent other profiles (projections) of this principal in
         // other tenants. Those tokens will be 'sparse', meaning that their AT/RT will not be loaded
         final ICacheRecord fullCacheRecord = cacheRecords.get(0);
+        if (fullCacheRecord.getAccessToken().refreshOnIsActive()) {
+            Logger.info(
+                    TAG,
+                    "RefreshOn is active. This will extend your token usage in the rare case servers are not available."
+            );
+        }
 
-        if (accessTokenIsNull(fullCacheRecord)
+        if(fullCacheRecord.getAccessToken().shouldRefresh()){
+            setAcquireTokenResult(acquireTokenSilentResult, parametersWithScopes, cacheRecords);
+            final RefreshOnCommand refreshOnCommand = new RefreshOnCommand(parameters, this, PublicApiId.MSAL_REFRESH_ON);
+            CommandDispatcher.submitAndForget(refreshOnCommand);
```

Comment:
Seems very strange...it looks like all you are doing here is doing the refresh and returning? Don't we need to proceed with the current request and return that token at least if it is not expired?

Reply:
That is what we are doing.

Reply:
Okay, got it it, so you return the result. How do you have guarantee that the access token you are returning is valid?

Reply:
The check to check if AT is expired happens further down the road

Reply:
Good catch.

Reply:
I've added logic to handle token expiry.